### PR TITLE
[exec] Fix %2$s exception

### PR DESCRIPTION
**Pull request by CameronDevine**
This is a Bugfix because of the issue with the Exec binding where using `%2$s` causes errors such as,
```
2017-08-15 20:07:12.902 [ERROR] [hab.binding.exec.handler.ExecHandler] - An exception occurred while formatting the command line with the current time and input values : 'Format specifier '%2$s''
```
This issue is referenced in #2605 and also in,

https://community.openhab.org/t/an-exception-occurred-while-formatting-the-command-line/33385
https://community.openhab.org/t/exec-binding-2-s-format-exception/33751/2

among other places.

The issue turned out to be in the `handleCommand()` function. The line `if (command instanceof StringType)` evaluates to false even when the input is from a `String` item. Once this if condition was removed the code acted as expected.

A prebuilt JAR of the modified Exec binding is included below.

[org.openhab.binding.exec-2.2.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab2-addons/files/1276255/org.openhab.binding.exec-2.2.0-SNAPSHOT.jar.zip)

### CODE REVIEW

**Comment by CameronDevine** [comment]
I forgot to sign off on one commit, and attempted to fix it, but it does not seem to have worked. Any tips on how to fix it would be appreciated.

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 1]
```diff
@@ -80,16 +80,15 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
-                    String previousInput = lastInput;
-                    lastInput = command.toString();
-                    if (lastInput != null && !lastInput.equals(previousInput)) {
-                        if (getConfig().get(AUTORUN) != null && ((Boolean) getConfig().get(AUTORUN)).booleanValue()) {
-                            lastInput = command.toString();
-                            logger.trace("Executing command '{}' after a change of the input channel to '{}'",
-                                    getConfig().get(COMMAND), command.toString());
-                            scheduler.schedule(periodicExecutionRunnable, 0, TimeUnit.SECONDS);
-                        }
+                logger.error("command = {}", command.toString());
```
Comment (Hilbrand) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: I think you forgot to remove this logging to error.

Comment (CameronDevine) [reply addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: I removed that line of code. An updated JAR file is located here,

[org.openhab.binding.exec-2.2.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab2-addons/files/1282574/org.openhab.binding.exec-2.2.0-SNAPSHOT.jar.zip)

**Comment by martinvw** [comment]
@kgoderis can you please check whether you agree that it should not check for the instanceOf.

@CameronDevine did you check what the actual type of the command was? It makes me curious what is going on, can you log for example `logger.info("Received {} with value {}", command.getClass(), command);`

**Comment by CameronDevine** [comment]
@martinvw I did check the type of the command while debugging the code. I was using a `Color` item as an input to my `exec` thing. It turned out the command was of the `HSBType` class. I didn't do further tests with other item types.

**Comment by martinvw** [comment]
@kgoderis wdyt?

**Comment by martinvw** [comment]
@kaikreuzer wdyt? Is this the proper fix? @kgoderis is most likely busy at this moment

**Comment by kaikreuzer** [comment]
> The line if (command instanceof StringType) evaluates to false even when the input is from a String item.

> It turned out the command was of the HSBType class.

These two statements contradict each other, right? I assume the first one CANNOT happen, the existing code should handle that just fine.

The second one is an illegal use case, because the "input" channel type must only be bound to String items - that's how it is defined.

**Comment by CameronDevine** [comment]
@kaikreuzer 
I did a bit more testing today. Using a String item as input does make `(command instanceof StringType)` evaluate to true. However using items such as Color as input makes `(command instanceof StringType)` evaluate to false, and therefore no command is run. Is this the desired result of this code?

**Comment by kaikreuzer** [comment]
>  Is this the desired result of this code?

Yes, because as I said above: The "input" channel is defined for "String" items, so it must not be linked to any other type.

**Comment by CameronDevine** [comment]
I have my code running with a HSBType input which I am using as input to a python script to convert to RGB values and sending them to a esp8266 over MQTT. What is the suggested way to accomplish this task if string inputs to the exec binding as the only ones allowed?

**Comment by halmafest** [comment]
Hi all, this issue prevents me at the moment to migrate from openhab1 to obenhab2. 

Part of my setup is a max7311 (i2c) which controls some relays. I have a small script which executes the required commands which enables me to switch on / off some lights.

To control my lights I created a switch item and connected it to the exec binding (Input) calling my script. It took me a while to figure out why this isn't working.  As the switch isn't a "String" items all the input is ignored. The execution isn't started.

As a workaround I created a String Item which acts as an intermediate state container. Instead of the switch I connected the String item to input channel of the exec binding and I wrote 2 rules to trigger the string item to be changed when the switch is used.

means : Switch [ON] triggers a rule to be run which changes the String Item to the text 'ON'. The binding takes it as valid input and executes the command using the string value 'ON'. A second rule is triggered when the Switch is set to OFF, String item text will be set to 'OFF' which triggers the exec binding to run my script switching off my lights. 
To get it working I have to define a useless String item and 2 rules. This feels quite wrong. Maybe I missed some concepts from OH2 and therefore I missed the better solution. 

Unfortunately I have not only one relay which I want to control with a switch, I have some more. I'm not willing to mess up my rules and item definitions with this workaround.

@kaikreuzer Why is the input channel of the exec binding limited to a String item? It does not really make sense to me as I expect every item type has implemented a toString() method producing in the most cases useful values.

**Comment by Hilbrand** [comment]
@halmafest Each channel has a type. The input channel of the exec binding is of type String. Therefor it is limited to use it only with a string.

In general if you look at the exec binding it seems to be specifically developed to run a command at a regular period and show the output as text. Like the state of a device. This doesn't seem to match with the requirement to use to send control commands to devices. This would require some additional thing/channels. I can give it a shot, but we probably should start with an issue (existing or new) or community thread to discuss the specifications.

**Pull request closed without merging**